java-squid/src/main/java/org/sonar/java/se/ExplodedGraphWalker.java

```diff
@@ -0,0 +1,377 @@
+          if (forStatement.condition() != null) {
+            handleBranch(block, forStatement.condition(), false);
+            return;
+          }
+      }
+    }
+    // unconditional jumps, for-statement, switch-statement:
+    for (CFG.Block successor : block.successors()) {
+      enqueue(new ExplodedGraph.ProgramPoint(successor, 0), programState);
+    }
+  }
+
+  private void handleBranch(CFG.Block programPosition, Tree condition) {
+    handleBranch(programPosition, condition, true);
+  }
+
+  private void handleBranch(CFG.Block programPosition, Tree condition, boolean checkPath) {
+    Pair<ProgramState, ProgramState> pair = constraintManager.assumeDual(programState, condition);
+    if (pair.a != null) {
+      // enqueue false-branch, if feasible
+      enqueue(new ExplodedGraph.ProgramPoint(programPosition.falseBlock(), 0), pair.a);
+      if (checkPath) {
+        alwaysTrueOrFalseChecker.evaluatedToFalse(condition);
+      }
+    } else {
+      SymbolicValue val = getVal(condition);
+      if (val != null) {
+        programState = ConstraintManager.setConstraint(programState, val, ConstraintManager.BooleanConstraint.TRUE);
+      }
+    }
+    if (pair.b != null) {
+      // enqueue true-branch, if feasible
+      enqueue(new ExplodedGraph.ProgramPoint(programPosition.trueBlock(), 0), pair.b);
+      if (checkPath) {
+        alwaysTrueOrFalseChecker.evaluatedToTrue(condition);
+      }
+    } else {
+      SymbolicValue val = getVal(condition);
+      if (val != null) {
+        programState = ConstraintManager.setConstraint(programState, val, ConstraintManager.BooleanConstraint.FALSE);
+      }
+    }
+
+  }
+
+  private void visit(Tree tree, Tree terminator) {
+    LOG.debug("visiting node " + tree.kind().name() + " at line " + ((JavaTree) tree).getLine());
+    switch (tree.kind()) {
+      case LABELED_STATEMENT:
+      case SWITCH_STATEMENT:
+      case EXPRESSION_STATEMENT:
+      case PARENTHESIZED_EXPRESSION:
+        throw new IllegalStateException("Cannot appear in CFG: " + tree.kind().name());
+      case VARIABLE:
+        VariableTree variableTree = (VariableTree) tree;
+        ExpressionTree initializer = variableTree.initializer();
+        if (initializer == null) {
+          if (terminator != null && terminator.is(Tree.Kind.FOR_EACH_STATEMENT)) {
+            setSymbolicValueForEachValue(variableTree);
+          } else if (variableTree.type().symbolType().is("boolean")) {
+            setSymbolicValueFalseValue(variableTree);
+          } else if (!variableTree.type().symbolType().isPrimitive()) {
+            setSymbolicValueNullValue(variableTree);
+          }
+        } else {
+          setSymbolicValueValueFromInitializer(variableTree, initializer);
+        }
+        break;
+      case ASSIGNMENT:
+        AssignmentExpressionTree assignmentExpressionTree = ((AssignmentExpressionTree) tree);
+        // FIXME restricted to identifiers for now.
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Take the required action to fix the issue indicated by this comment. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.internal.sonarsource.com/coding_rules#rule_key=squid%3AS1134)